### PR TITLE
Open up v2 http status code checks for put and head checks

### LINKED ISSUE #10402: http status code checks in registry/session_v2.go too narrow
[Checks](https://github.com/docker/docker/blob/master/registry/session_v2.go#L158) that are [too narrow](https://github.com/docker/docker/blob/master/registry/session_v2.go#L245) are not compliant with the V2 specification. All 2xx and 3xx responses should generally be treated as successful. The body should only be decoded as an error body on 4xx and 5xx responses. These narrow checks will prevent future protocol evolution if a different error code is used in particular cases. It will also cause issues with proxies that may rewrite responses based on certain behavior.

A salient example may be returning a 304 instead of a 202 when putting a manifest. The method should only be concerned that the content safely made it, with a success http code. Whether it is a 202 or 304 or 307 (won't actually be propagated with go http client) doesn't really matter.

This does not apply when the method requires data to be returned but the error must be handled differently.

Please see the following for more information: https://github.com/docker/distribution/blob/master/doc/SPEC.md#errors-1

### PR DESCRIPTION
**Pull request by stevvooe**
Under certain cases, such as when putting a manifest or check for the existence
of a layer, the status code checks in session_v2.go were too narrow for their
purpose. In the case of putting a manifest, the handler only cares that an
error is not returned. Whether it is a 304 or 202 does not matter, as long as
the server reports success. Having the client only accept specific http codes
inhibits future protocol evolution.

Signed-off-by: Stephen J Day stephen.day@docker.com

Closes #10402.

### CODE REVIEW

**Comment by stevvooe** [comment]
cc @icecrime @jfrazelle

File: registry/session_v2.go [hunk 1]
```diff
@@ -124,14 +124,15 @@ func (r *Session) HeadV2ImageBlob(ep *Endpoint, imageName, sumType, sum string,
 		return false, err
 	}
 	res.Body.Close() // close early, since we're not needing a body on this call .. yet?
-	switch res.StatusCode {
-	case 200:
+	switch {
```
Comment (LK4D4) [inline registry/session_v2.go]: should be just two `if`s

File: registry/session_v2.go [hunk 2]
```diff
@@ -278,7 +279,9 @@ func (r *Session) PutV2ImageManifest(ep *Endpoint, imageName, tagName string, ma
 		return err
 	}
 	defer res.Body.Close()
-	if res.StatusCode != 200 {
+
+	// All 2xx and 3xx responses can be accepted for a put.
+	if res.StatusCode >= 400 {
```
Comment (LK4D4) [inline registry/session_v2.go]: Should be
```
if res.StatusCode < 400 {
    return nil
}
```

File: registry/session_v2.go (hunk 1 above)
Comment (stevvooe) [inline registry/session_v2.go]: `switch` implies coverage over an input domain, which is what we want here. Arguably, default return should be pulled up into switch for clarity.

File: registry/session_v2.go (hunk 2 above)
Comment (stevvooe) [inline registry/session_v2.go]: Success is part of main function flow whereas as errors are branches.

File: registry/session_v2.go [hunk 3]
```diff
@@ -124,14 +124,15 @@ func (r *Session) HeadV2ImageBlob(ep *Endpoint, imageName, sumType, sum string,
 		return false, err
 	}
 	res.Body.Close() // close early, since we're not needing a body on this call .. yet?
-	switch res.StatusCode {
-	case 200:
+	switch {
+	case res.StatusCode >= 200 && res.StatusCode < 400:
```
Comment (icecrime) [inline registry/session_v2.go]: I find it a bit counterintuitive that things outside 200 is interpreted as success. Can you explain a bit more? How for example would a 301 be understood as "no push needed"?

File: registry/session_v2.go (hunk 1 above)
Comment (LK4D4) [inline registry/session_v2.go]: You're not covering input domain of http status codes here and this looks just weird.

File: registry/session_v2.go (hunk 3 above)
Comment (stevvooe) [inline registry/session_v2.go]: It's simply that a 301 is not an error. If a 2xx or 3xx response is issued, it should be assumed that the content is known to the registry and *does not need to be pushed*. Whether or not its actually available at that location doesn't matter. While the registry may never return this error code, the client should assume that it's a success and continue.

In reality, 301 should be handled transparently by the Go http client, resulting in an eventual 200 or 404 response. However, opening up the accepted error codes here allows us to modify the protocol and maintain backwards compatibility with 1.5.

Comment (icecrime) [inline registry/session_v2.go]: Got it, thanks!

**Comment by icecrime** [comment]
This sounds reasonable to me, and has very low impact. I'm not going to take position in the style comments, so LGTM!

**Comment by icecrime** [comment]
@LK4D4 Do we have your LGTM or do you feel your comments are blockers?

**Comment by LK4D4** [comment]
LGTM
Though I'm not best person to ask :)

**Comment by tiborvass** [comment]
LGTM

**Pull request merged**